core/src/main/java/org/infinispan/commands/irac/IracTombstoneRemoteSiteCheckCommand.java

```diff
@@ -62,18 +87,42 @@ public byte getCommandId() {
    }
 
    @Override
-   public CompletionStage<Boolean> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
-      DistributionInfo distribution = registry.getDistributionManager().getCacheTopology().getDistribution(key);
-      if (distribution.isPrimary()) {
-         return isKeyInIracManager(registry);
-      } else {
-         RpcManager manager = registry.getRpcManager().running();
-         return manager.invokeCommand(distribution.primary(), this, new BooleanResponseCollector(), manager.getSyncRpcOptions());
+   public CompletionStage<IntSet> performInLocalSite(ComponentRegistry registry, boolean preserveOrder) {
+      LocalizedCacheTopology topology = registry.getDistributionManager().getCacheTopology();
+      IracManager iracManager = registry.getIracManager().running();
+      RpcManager rpcManager = registry.getRpcManager().running();
+      RpcOptions rpcOptions = rpcManager.getSyncRpcOptions();
+      int numberOfKeys = keys.size();
+
+      Map<Address, IntSetResponseCollector> primaryOwnerKeys = new HashMap<>(rpcManager.getMembers().size());
+      IntSet toKeepIndexes = IntSets.concurrentSet(numberOfKeys);
+
+      for (int index = 0 ; index  < numberOfKeys; ++index) {
+         Object key = keys.get(index);
+         DistributionInfo dInfo = topology.getDistribution(key);
```

Comment:
The only other thing I can think of to help is possibly sending the keys along with what segment they are. We are computing the segment twice, but tbh it is probably pretty minor. Unfortunately it would also increase the message size ever so slightly (based on unique segments), so I am fine integrating as is.